NEW: add environment loading and dynamic settings

streeck opened the pull request.

streeck committed:
NEW: add environment loading and dynamic settings

kasterma commented on tox.ini:
```diff
@@ -2,17 +2,18 @@
     coverage xml -i
 deps =
-    -rrequirements/requirements-base.txt
+    # -rrequirements/requirements-base.txt
```
b/c empty?

streeck replied:
Yeah, it was throwing some errors with tox. For now this will be empty, so maybe we should remove this line + the file 🤔

kasterma commented on settings_vial/settings_vial.py:
```diff
@@ -0,0 +1,110 @@
+                try:
+                    json_value = json.loads(value)
+                except json.decoder.JSONDecodeError:
+                    json_value = value
```
remind: comment on why this for strings.

kasterma commented on settings_vial/__init__.py:
```diff
@@ -0,0 +1,7 @@
+
+__all__ = [Settings]
+
+logging.getLogger(__name__).addHandler(logging.NullHandler())
```
I don't think this does what you think it is doing.  It turns off default logging to stdout/stderr (https://docs.python.org/3/howto/logging.html#configuring-logging-for-a-library).  Also b/c you don't use the logger, but the warning module, the relevant logger would be 'py.warnings' (https://docs.python.org/3.7/library/warnings.html, and https://docs.python.org/3.7/library/logging.html#logging.captureWarnings)

kasterma replied:
Actually, since I can't get captureWarnings to work it is clear I still don't understand something here.

streeck replied:
You're absolutely right, I chose my words poorly. What I meant was that the library should not handle log events, the user should decide how those are handled. By adding the `NullHandler` as you mentioned, it then does nothing about those events.
By using `__name__` we ensure no naming collisions with the loggers.

From this link: https://docs.python.org/3/howto/logging.html#configuring-logging-for-a-library I think the most relevant part is the **Note** box.

An example usage so that you can see the `logging.captureWarnings(True)` working:

```python
import logging
from settings_vial import Settings

logging.captureWarnings(True)
logging.basicConfig(filename='app.log', filemode='w', format='%(name)s - %(levelname)s - %(message)s')

settings = Settings('PREFIX_', 'OVERRIDE_', 'not_callable')
settings.BLA
```
You can read more about the different attributes in the basicConfig format here: https://docs.python.org/3/library/logging.html#logrecord-attributes

This would write the following to `app.log`:
```
py.warnings - WARNING - .../settings-vial/settings_vial/settings_vial.py:67: UnsupportedSetType: Override callable does not return a set or a list
  warnings.warn("Override callable does not return a set or a list", UnsupportedSetType)
```

kasterma replied:
Problem with my testing of `warning` module was that I hadn't read the docs well enough.  Repeat warnings are ignored (https://docs.python.org/3.7/library/warnings.html#the-warnings-filter).

kasterma commented on settings_vial/settings_vial.py:
```diff
@@ -0,0 +1,110 @@
+        42
+    """
+
+    def __init__(self, env_prefix, override_prefix=None, override_callable=None):
```
Not certain about the actual use of override_callable; I see the tests and that they pass.

streeck replied:
So, think of the following scenario, I have a certain setting for enabling throttling in certain endpoints, by default it's turned off but in my e2e tests I want to enable it, but only for a certain set of my tests/requests.
Passing a callable (e.g. a function that fetches specific request headers) that returns the key(s) you defined as your override settings, it will use those alternative settings for that request. It is mainly aimed at e2e tests.

ricardosantosalves replied:
I think `override_callable` is not a good name. What about `override_keys_enumerator` or `override_keys_function`?

kasterma reviewed (commented):
tests/__init__.py rename; maybe just delete the file?

streeck commented:
> tests/**init**.py rename; maybe just delete the file?

I did delete it, but believe `git` tried to be smart about it 😓

ricardosantosalves commented on settings_vial/exceptions.py:
```diff
@@ -0,0 +1,16 @@
+class NotCallable(Warning):
```
minor (naming): for clarity, I think these exceptions should suffixed with `Exception` or `Error`.

streeck replied:
Maybe `Warning` then?

ricardosantosalves commented on settings_vial/settings_vial.py:
```diff
@@ -0,0 +1,110 @@
+
+    :param env_prefix: prefix to be used when looking through the environment for variables
+    :param override_prefix: prefix to be used when looking for overrideable variables
+    :param override_callable: callable to be used when retrieving the set of keys to be overriden
```
Maybe this could be something like: "callable returning a set of keys to be used as overrides."

streeck replied:
I'll go for `override_keys_function`, I agree it's more intuitive.
Also updated the docstring

jeduden commented on setup.py:
```diff
@@ -0,0 +1,38 @@
+    zip_safe=False,
+    license="Apache 2.0",
+    classifiers=[
+        "Development Status :: 3 - Alpha",
```
shoudn't the status be stable ?

streeck replied:
Well, there were no releases so far, I'd like to see how it goes first 😅

jeduden replied:
so when we run this in prod will you update the label ? 😎

streeck replied:
Absolutely! 😎

jeduden commented on tests/test_settings_vial.py:
```diff
@@ -1,19 +1,141 @@
+        with pytest.raises(AttributeError):
+            settings.USER
+        with pytest.raises(AttributeError):
+            settings.PREFIX_DEBUG
```
why shoud accessing PREFIX_DEBUG result in an AttributeError ?

streeck replied:
Because the settings will be accessible through `settings.DEBUG` since the `PREFIX` is stripped out of it during the load

jeduden replied:
Ok. Have you considered to move this assertion to its own test so that you can better explain this expectation ?

streeck replied:
That would indeed be a better option, I'll move it later 😉

streeck committed:
FIX: rewrite travis.yml

Removed linting dependencies from requirements_testing to tox.ini
Add the {toxinidir} as editable dependency in tox.ini

streeck committed:
FIX: replace setup_gen with setup.py and update gitignore

streeck committed:
OPT(refactor): add Warning suffix to all warnings

streeck committed:
OPT: update travis with deploy key and docs

streeck committed:
OPT(refactor): rename override_callable parameter

streeck committed:
OPT: remove setup_gen leftovers from Makefile
